rust/src/router.rs

```diff
@@ -1,38 +1,90 @@
+use crate::tree::RadixTree;
 use actix_web::http::header::{HeaderValue, CONTENT_TYPE};
 use actix_web::{HttpRequest, HttpResponse};
 use bytes::Bytes;
 use futures_util::TryStreamExt;
+use std::collections::HashMap;
 use std::fmt::Debug;
+use std::sync::atomic::AtomicUsize;
+use std::sync::{Arc, Mutex};
+use tokenizers::tokenizer::Tokenizer;
 
 #[derive(Debug)]
 pub enum Router {
     RoundRobin {
         worker_urls: Vec<String>,
-        current_index: std::sync::atomic::AtomicUsize,
+        current_index: AtomicUsize,
     },
     Random {
         worker_urls: Vec<String>,
     },
+    ApproxTree {
+        worker_urls: Vec<String>,
+        // TODO: don't lock the whole tree
+        url_to_tree: Arc<Mutex<HashMap<String, RadixTree>>>,
```

Comment:
[Dashmap](https://github.com/xacrimon/dashmap) may helps here

Reply:
Actually i want to lock on the node of radixtree instead of the whole hashmap. Do you have any suggestion for that?

Reply:
I think we can start from the simple one (lock whole map). Concurrent map is quite complex. And may try out this comment later on if you are interested
https://users.rust-lang.org/t/locking-only-one-entry-in-hashmap/84764/4